### ISSUE TITLE
[ENH] Make function pointer exception specification matching more lenient

### ISSUE THREAD

**Issue by da-woods** [cython/cython#4689]
**Is your feature request related to a problem? Please describe.**

This relates to https://github.com/cython/cython/issues/4280 (where it's unclear what exception specification should be assumed for function pointers). In my mind a `noexcept` function is compatible with a pointer with an `except *` or an `except ?value` pointer specification (but not an `except value` or `except +` specification). This is because it simply adds an unnecessary check that always passes.

Similarly an `except value` function should be compatible with `except *` and an `except ?value` specification. (And a few other similar combinations)

Code to demonstrate (that doesn't currently work):

```cython
cdef extern from *:
    cdef int extern_f()  # noexcept

cdef int takes_func(int (*f)() except *):
    return f()

def call():
    takes_func(extern_f)  # fails, but shouldn't
```

**Additional context**

This is consistent with what C++ does (which doesn't necessarily say anything about what Cython should do...)
```c++
int noexcept_func() noexcept;
int regular_func();

int (*f)() = noexcept_func;  // OK
int (*g)() = regular_func; // OK
int (*h)() noexcept = noexcept_func; // OK
int (*i)() noexcept = regular_func; // fails
```

The advantage to this is that we could assume `except *` for function pointers by default and the vast majority of code would keep working.

I think it's probably worth doing independently whatever we decide on https://github.com/cython/cython/issues/4280 (and thus I create a separate issue)

**Comment by scoder** [comment]
Seems the right thing to do, yes.

**Comment by da-woods** [comment]
FWIW this partially seems to work - I'm investigating where it doesn't match. There may be a specific issue with ints more than the general principle.

**Comment by da-woods** [comment]
I can't easily find any tests for this, but I think it generally is implemented (and I was being mislead by #4770). I'll try to add some tests but I no longer think this is blocking anything.
